resources/net.go

```diff
@@ -0,0 +1,804 @@
+	contents := obj.unitFileContents()
+	// check the file contents
+	if err == nil {
+		unitFile, err := ioutil.ReadFile(obj.unitFilePath)
+		if err != nil {
+			return false, errwrap.Wrapf(err, "error reading file")
+		}
+		// return if the file is good
+		if bytes.Equal(unitFile, contents) {
+			return true, nil
+		}
+	}
+
+	if !apply {
+		return false, nil
+	}
+
+	log.Printf("%s: fileCheckApply(%t)", obj, apply)
+
+	// write the file
+	if err := ioutil.WriteFile(obj.unitFilePath, contents, 0644); err != nil {
+		return false, errwrap.Wrapf(err, "error writing configuration file")
+	}
+	return false, nil
+}
+
+// ipCheckApply checks if the interface has the correct addresses and restarts
+// the networkd service if the addresses are inconsistent.
+func (obj *NetRes) ipCheckApply(apply bool) (checkOK bool, err error) {
+	var addrStrings []string
+	addrOK := true
+
+	if obj.Addrs == nil {
+		return true, nil
+	}
+
+	// if the interface is configuring, wait for the next event
+	// create a bus object pointing to the adapter
+	busObj := obj.gdbus.Object(dBusNetworkdInterface, obj.ifacePath)
+	// get the object's operational state
+	call := busObj.Call(dBusPropertiesGet, 0, dBusNetworkdLinkInterface, adminState)
+	if call.Err != nil {
+		return false, call.Err
+	}
+	// check the operational state
+	as, ok := call.Body[0].(dbus.Variant)
+	if !ok {
+		return false, fmt.Errorf("error casting opstate")
+	}
+	if as == dbus.MakeVariant(adminStateConfiguring) {
+		return false, nil
+	}
+
+	// find the specified interface and get a list of addresses
+	iface, err := net.InterfaceByName(obj.GetName())
+	if err != nil {
+		return false, errwrap.Wrapf(err, "error finding interface: %s", obj.GetName())
+	}
+	addrs, err := iface.Addrs()
+	if err != nil {
+		return false, errwrap.Wrapf(err, "error getting addrs from interface: %s", obj.GetName())
+	}
+	// if the address is not a link local address, add it to the list
+	for _, addr := range addrs {
+		ip, _, err := net.ParseCIDR(addr.String())
+		if err != nil {
+			return false, errwrap.Wrapf(err, "error parsing interface addr: %s", addr)
+		}
+		// XXX: ignore multicast too?
+		if !ip.IsLinkLocalUnicast() {
+			addrStrings = append(addrStrings, addr.String())
+		}
+	}
+
+	// compare defined addrs with actual addrs
+	objAddrs := obj.Addrs
+	sort.Strings(objAddrs)
+	sort.Strings(addrStrings)
+	if len(objAddrs) == len(addrStrings) {
+		for i := range obj.Addrs {
+			if obj.Addrs[i] != addrStrings[i] {
+				addrOK = false
+			}
+		}
+		// if the addresses check out, we're done
+		if addrOK {
+			return true, nil
+		}
+	}
+
+	// end of state checking. if we're here, checkOK is false
+	if !apply {
+		return false, nil
+	}
+
+	log.Printf("%s: ipCheckApply(%t)", obj, apply)
+
+	// restart networkd to re-initialize the interface
+	if err := restartSystemdUnit(obj.sdbus, networkdService, dBusFailMode); err != nil {
+		return false, errwrap.Wrapf(err, "error restarting networkd service")
+	}
+
```

Comment:
do we need to wait here? what happens with the restartSystemdUnit stuff?
please include your findings about how it works as a comment :)